### PR TITLE
Update pydantic qcmetrics

### PR DESCRIPTION
**Pull request by henrikstranneheim**
### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b update-pydantic-qcmetrics
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
### cg workflow balsamic start sweetelf --dry-run, 👍 
```
henrik.stranneheim@hasta:~] [S_base] $ cg workflow balsamic start sweetelf --dry-run
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L001.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L004.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L002.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L003.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L001.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L004.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L002.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L003.crunchy.pending.txt does not exist
Compression/decompression is not running
Both /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L001_R1_001.fastq.gz and /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L001_R2_001.fastq.gz from sample ACC6637A1 are already in Housekeeper
Both /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L004_R1_001.fastq.gz and /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L004_R2_001.fastq.gz from sample ACC6637A1 are already in Housekeeper
Both /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L002_R1_001.fastq.gz and /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L002_R2_001.fastq.gz from sample ACC6637A1 are already in Housekeeper
Both /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L003_R1_001.fastq.gz and /home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L003_R2_001.fastq.gz from sample ACC6637A1 are already in Housekeeper
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L001.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L004.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L002.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L003.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if FASTQ pair exists
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L001.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L004.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L002.crunchy.pending.txt does not exist
Compression/decompression is not running
Check if pending compression file exists
/home/proj/stage/housekeeper-bundles/ACC6637A1/2020-05-10/H525HDSXY_GMSmyeloid_S171_L003.crunchy.pending.txt does not exist
Compression/decompression is not running
Starting analysis for sweetelf
Case sweetelf exists in Status DB
Creating config file for sweetelf.
Case sweetelf exists in Status DB
Case sweetelf has following BALSAMIC samples:
SAMPLE ID            TISSUE TYPE          APPLICATION          BED VERSION
ACC6637A1            tumor                tgs                  gmsmyeloid_5.2_hg19_design.bed

The provided sex is unknown, setting female as the default
Running command /home/proj/stage/bin/miniconda3/bin/conda run --name S_balsamic /home/proj/stage/bin/miniconda3/envs/S_balsamic/bin/balsamic config case --analysis-dir /home/proj/stage/cancer/cases --analysis-workflow balsamic --balsamic-cache /home/proj/stage/cancer/balsamic_cache/ --cadd-annotations /home/proj/stage/cancer/reference/cadd/data/annotations --cancer-germline-snv-observations /home/proj/stage/cancer/reference/loqusdb/loqusdb_cancer_germline_snv_variants_export-20241017-.vcf.gz --cancer-somatic-snv-observations /home/proj/stage/cancer/reference/loqusdb/loqusdb_cancer_somatic_snv_variants_export-20241017-.vcf.gz --cancer-somatic-sv-observations /home/proj/stage/cancer/reference/loqusdb/loqusdb_cancer_somatic_sv_variants_export-20241017-.vcf.gz --case-id sweetelf --clinical-snv-observations /home/proj/stage/cancer/reference/loqusdb/loqusdb_clinical_snv_variants_export-20241017-.vcf.gz --clinical-sv-observations /home/proj/stage/cancer/reference/loqusdb/loqusdb_clinical_sv_variants_export-20241017-.vcf.gz --fastq-path /home/proj/stage/cancer/cases/sweetelf/fastq --gender female --genome-version hg19 --panel-bed /home/proj/stage/cancer/reference/panel/gmsmyeloid_5.2_hg19_design.bed --swegen-snv /home/proj/stage/cancer/reference/swegen/grch37_swegen_1k_snv_-20170830-.vcf.gz --swegen-sv /home/proj/stage/cancer/reference/swegen/grch37_swegen_1k_sv_-20170830-.vcf.gz --tumor-sample-name ACC6637A1
Dry run: process call will not be executed!!
Case sweetelf exists in Status DB
Running command /home/proj/stage/bin/miniconda3/bin/conda run --name S_balsamic /home/proj/stage/bin/miniconda3/envs/S_balsamic/bin/balsamic run analysis --account development --mail-user clinical.delivery.cancer@scilifelab.se --qos normal --sample-config /home/proj/stage/cancer/cases/sweetelf/sweetelf.json --benchmark
Dry run: process call will not be executed!!

File: cg/models/balsamic/config.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 from datetime import datetime
 from pathlib import Path
 
-from pydantic.v1 import BaseModel, validator
+from pydantic import BaseModel, field_validator
+from pydantic_core.core_schema import ValidationInfo
```
Comment (diitaz93) [inline cg/models/balsamic/config.py:5]: You could simplify this import by importing ValidationInfo directly from pydantic
```suggestion
from pydantic import BaseModel, field_validator, ValidationInfo
```

**Review by diitaz93 — APPROVED** [review]
Looks great 💯

File: cg/models/balsamic/metrics.py [hunk 2]
```diff
@@ -17,60 +17,56 @@ class BalsamicMetricsBase(MetricsBase):
...
-    pct_pf_reads_improper_pairs: float | None
+    median_coverage: float | None = None
+    pct_15x: float | None = None
+    pct_30x: float | None = None
+    pct_60x: float | None = None
+    pct_100x: float | None = None
+    pct_pf_reads_improper_pairs: float | None = None
 
-    _pct_values = validator(
+    _pct_values = field_validator(
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply cg/models/balsamic/metrics.py]: Ficed

File: cg/models/balsamic/metrics.py [hunk 3]
```diff
@@ -17,60 +17,56 @@ class BalsamicMetricsBase(MetricsBase):
...
+    median_target_coverage: float | None = None
+    pct_target_bases_50x: float | None = None
+    pct_target_bases_100x: float | None = None
+    pct_target_bases_250x: float | None = None
+    pct_target_bases_500x: float | None = None
+    pct_target_bases_1000x: float | None = None
+    pct_off_bait: float | None = None
+    gc_dropout: float | None = None
+
+    _pct_values = field_validator(
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply cg/models/balsamic/metrics.py]: Fixed

File: cg/models/balsamic/metrics.py [hunk 4]
```diff
@@ -17,60 +17,56 @@ class BalsamicMetricsBase(MetricsBase):
...
-    mean_insert_size: float | None
-    percent_duplication: float | None
+    fold_80_base_penalty: float | None = None
+    mean_insert_size: float | None = None
+    percent_duplication: float | None = None
 
-    _percent_duplication = validator("percent_duplication", allow_reuse=True)(
-        percent_value_validation
-    )
+    _percent_duplication = field_validator("percent_duplication")(percent_value_validation)
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply cg/models/balsamic/metrics.py]: Fixed

**Comment by henrikstranneheim** [comment]
se@Clinical-Genomics/sysdev Thx for reverting! Any error or cmd/case that I can use to test this error?

**Comment by islean** [comment]
> @Clinical-Genomics/sysdev Thx for reverting! Any error or cmd/case that I can use to test this error?

No worries! I was just about to ping you with some info! Generating delivery reports for Balsamic cases started crashing (and by extension uploads). My hunch is that it has to do with the "always=True" behaviour changing in v2 since it seemed as if the capture kit version and genome builds were not set. I tried fixing it but we had some time critical cases that needed uploading and the Balsamic test environment is seeing some changes due to the new release so testing was slow.

**Pull request merged**